corehq/apps/aggregate_ucrs/ingestion.py

```diff
@@ -0,0 +1,186 @@
+"""
+This module deals with data ingestion: populating the aggregate tables from other tables.
+"""
+from __future__ import absolute_import, unicode_literals
+from collections import namedtuple
+from datetime import datetime
+import sqlalchemy
+from sqlalchemy.dialects.postgresql import insert
+
+from corehq.apps.aggregate_ucrs.aggregations import AGG_WINDOW_START_PARAM, AGG_WINDOW_END_PARAM, \
+    TimeAggregationWindow
+from corehq.apps.userreports.sql import IndicatorSqlAdapter
+from dimagi.utils.parsing import json_format_date
+
+
+AggregationParam = namedtuple('AggregationParam', 'name value mapped_column_id')
+AggregationWindow = namedtuple('AggregationWindow', 'start end')
+
+
+def populate_aggregate_table_data(aggregate_table_adapter):
+    """
+    Seeds the database table with all data from the table adapter.
+    """
+    aggregate_table_definition = aggregate_table_adapter.config
+    # get checkpoint
+    last_update = get_last_aggregate_checkpoint(aggregate_table_definition)
+    for window in get_time_aggregation_windows(aggregate_table_definition, last_update):
+        populate_aggregate_table_data_for_time_period(
+            aggregate_table_adapter, window,
+        )
+
+
+def get_last_aggregate_checkpoint(aggregate_table_definition):
+    """
+    Checkpoints indicate the last time the aggregation script successfully ran.
+    Will be used to do partial ingestion.
+    """
+    # todo
+    return None
+
+
+def get_time_aggregation_windows(aggregate_table_definition, last_update):
+    if aggregate_table_definition.time_aggregation is None:
+        # if there is no time aggregation just include a single window with no value
+        yield None
+    else:
+        start_time = get_aggregation_start_period(aggregate_table_definition, last_update)
+        end_time = get_aggregation_end_period(aggregate_table_definition, last_update)
+        assert end_time >= start_time
+        adapter_class = TimeAggregationWindow.from_aggregation_unit(
+            aggregate_table_definition.time_aggregation.aggregation_unit
+        )
+        current_window = adapter_class(start_time)
+        end_window = adapter_class(end_time)
+        while current_window <= end_window:
+            yield AggregationWindow(
+                start=AggregationParam(
+                    name=AGG_WINDOW_START_PARAM,
+                    value=json_format_date(current_window.start),
+                    mapped_column_id=aggregate_table_definition.time_aggregation.start_column
+                ),
+                end=AggregationParam(
+                    name=AGG_WINDOW_END_PARAM,
+                    value=json_format_date(current_window.end),
```

Comment:
JSON format could be done in the `AggregationParam` class

Reply:
I moved it to the `TimeAggregationWindow` class, though could make `AggregationParam` heavier if you think that's a better way to go